### PR TITLE
v4.0.3 Functionality

### LINKED ISSUE #252: User report: algorithm choosing wrong arms (correctness)
A report from a user, I've asked for his code:

> I compared the BUILD step solution's objective of BanditPAM with that of the BUILD step solution's objective of PAM (on MNIST dataset, train split, N = 60k, d = 784). For the number of medoids k = 100 and L2 distance, PAM's BUILD objective is 6.0641 while that of BanditPAM's BUILD is 6.55 (around 8% worse). I have observed such differences for various other randomly generated datasets as well. For the MNIST dataset, I had set build_confidence = 16. According to my understanding of BanditPAM's paper and code, this should correspond to delta = (N*exp(build_confidence))^{-1} = 1.8756e-12. Hence, according to Theorem 1's Remark A1 of BanditPAM paper, the probability that all confidence intervals are true confidence intervals is at least 1 - 2N^2*delta = 0.9932. My understanding is that this probability should be high enough for me to observe the same BUILD objective for both BanditPAM and PAM for most random seeds. However, it does not happen for any seed. 

> Please note that my understanding of the theoretical results or the way I used build_confidence may be wrong. Hence, it would be great if you could resolve the above issue. Overall, based on the results shown in the BanditPAM paper, I expected the objectives to be the same in most of my runs (on various datasets with 1 - 2N^2*delta > 0.99, computed through build_confidence as explained above). However, it does not happen and the objectives are quite far apart (e.g., 8% difference).

### LINKED ISSUE #256: Change github actions that upload wheels to upload to TestPyPI on PR update, and PyPI on release
Right now, both TestPyPI and PyPI actions trigger on release and we govern which one executes by commenting/uncommenting out the correct lines of code. We should automate this.

[Example](https://github.com/motiwari/BanditPAM/blob/284fea66e909d247b11c972b71efc82ce3e3319a/.github/workflows/build_linux_wheels.yml#L71-L78)

### PR DESCRIPTION
**Pull request by Adarsh321123**
BanditPAM `v4.0.3` contains the following _functionality_ changes:

**Organization and Functionality:**
- Ensure GHAs don't run on a tag push.
- Automatically upload wheels to TestPyPI on a PR update and PyPI on release in GHAs (fixes #256)
- Add `tests/test_initialization` to the GHAs that build and test the package.
- Loss issues are resolved in all versions of BanditPAM
- The clustering results are now identical to `scikit-learn`'s implementation. This was achieved by increasing the batch size for accurate estimation of the standard deviation of the arm parameters and by fixing the bug that drops arms whose lower confidence bounds equal the lowest upper confidence bounds (fixes #252).
- Add complexity to `scripts/comparison_utils.py`, such as printing cache writes.
- Update cache calculations in `src/algorithms/kmedoids_algorithm.cpp` and add an assertion for better error handling.
- Turn `VERSION_INFO` into a string in `src/python_bindings/kmedoids_pywrapper.cpp`.
- Update `build_confidence` in `src/python/kmedoids_pywrapper.cpp` to avoid stochasticity issue in `tests/test_larger.py`.

**Tests:**
- Update assertion in `tests/test_initialization` to match with new `build_confidence` in `src/python/kmedoids_pywrapper.cpp`.
- Use NumPy and Armadillo seeds for reproducibility.
- Update some tests in `tests/test_larger.py` to count the proportion of passing cases instead of immediately failing.

**Documentation:**
- Slightly updated the documentation on the CMake build on Windows to clarify how to use `retrieve_windows_cmake_files.sh`.

### CODE REVIEW

File: scripts/cache_measurements.py [hunk 1]
```diff
@@ -55,22 +55,30 @@ def test_cache_stats():
...
+        (
+            time_250,
+            width_250,
+            writes_250,
+            hits_250,
+            misses_250,
+        ) = get_cache_statistics(kmed=kmed, X=X, loss="L2", cache_width=250)
+        time_0, width_0, writes_0, hits_0, misses_0 = get_cache_statistics(
+            kmed=kmed, X=X, loss="L2", cache_width=0
+        )
```
Comment (lukeleeai) [inline scripts/cache_measurements.py:81]: Unify the format for consistency?

File: scripts/cache_measurements.py [hunk 2]
```diff
@@ -92,8 +100,7 @@ def test_cache_stats():
         assert width_250 == 250, "Cache width should be 250 when set to 250"
         assert width_500 == 500, "Cache width should be 500 when set to 500"
         assert width_750 == 750, "Cache width should be 750 when set to 750"
-        assert width_1000 == 1000, "Cache width should be 1000 when set to " \
-                                   "1000"
+        assert width_1000 == 1000, "Cache width should be 1000 when set to " "1000"
```
Comment (lukeleeai) [inline scripts/cache_measurements.py:103]: `when set to " "1000"` -> `when set to 1000"`

File: src/algorithms/banditpam.cpp [hunk 3]
```diff
@@ -104,6 +108,9 @@ namespace km {
       }
       updated_sigma(i) = arma::stddev(sample);
     }
+    // reset batchSize the original batch size as it's a global variable
```
Comment (lukeleeai) [inline src/algorithms/banditpam.cpp:111]: sorry I should have added "to" between "batchSize" and "the original batch size"

File: src/algorithms/banditpam.cpp [hunk 4]
```diff
@@ -319,6 +340,9 @@ namespace km {
       }
       updated_sigma(k, n) = arma::stddev(sample);
     }
+    // reset batchSize the original batch size as it's a global variable
```
Comment (lukeleeai) [inline src/algorithms/banditpam.cpp:343]: sorry I should have added "to" between "batchSize" and "the original batch size"

File: src/algorithms/banditpam_orig.cpp [hunk 5]
```diff
@@ -104,6 +108,9 @@ namespace km {
       }
       updated_sigma(i) = arma::stddev(sample);
     }
+    // reset batchSize the original batch size as it's a global variable
```
Comment (lukeleeai) [inline src/algorithms/banditpam_orig.cpp:111]: sorry I should have added "to" between "batchSize" and "the original batch size"

File: src/algorithms/banditpam_orig.cpp [hunk 6]
```diff
@@ -319,6 +340,9 @@ namespace km {
       }
       updated_sigma(k, n) = arma::stddev(sample);
     }
+    // reset batchSize the original batch size as it's a global variable
```
Comment (lukeleeai) [inline src/algorithms/banditpam_orig.cpp:343]: sorry I should have added "to" between "batchSize" and "the original batch size"

**Review by lukeleeai — COMMENTED** [review]
1. I will update my branch again and
2. open another PR from my branch to this 4.0.3 branch

File: tests/test_initialization.py [hunk 7]
```diff
@@ -78,8 +78,7 @@ def test_initialization_3(self):
         assert kmed.n_medoids == 100, "n_medoids does not match"
         assert kmed.algorithm == "PAM", "algorithm does not match"
         assert kmed.max_iter == 10000, "max_iter does not match"
-        assert kmed.build_confidence == 11000, "build_confidence does not " \
-                                               "match"
+        assert kmed.build_confidence == 11000, "build_confidence does not " "match"
```
Comment (lukeleeai) [inline tests/test_initialization.py:81]: remove " "

**Review by lukeleeai — CHANGES_REQUESTED** [review]
Fix one last minor typo
